packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,213 @@
+import { VM, scheduleRehydration } from './vm';
+import { isFalse, isUndefined, isNull } from '@lwc/shared';
+import { markComponentAsDirty, ComponentConstructor } from './component';
+import { Template } from './template';
+import { StylesheetFactory } from './stylesheet';
+
+const swappedTemplateMap: WeakMap<Template, Template> = new WeakMap();
+const swappedComponentMap: WeakMap<ComponentConstructor, ComponentConstructor> = new WeakMap();
+const swappedStyleMap: WeakMap<StylesheetFactory, StylesheetFactory> = new WeakMap();
+
+const activeTemplates: WeakMap<Template, Set<VM>> = new WeakMap();
+const activeComponents: WeakMap<ComponentConstructor, Set<VM>> = new WeakMap();
+const activeStyles: WeakMap<StylesheetFactory, Set<VM>> = new WeakMap();
+
+function rehydrateHotTemplate(tpl: Template) {
+    const list = activeTemplates.get(tpl);
+    list?.forEach((vm) => {
+        if (!isUndefined(vm)) {
+            if (isFalse(vm.isDirty)) {
+                // forcing the vm to rehydrate in the next tick
+                markComponentAsDirty(vm);
+                scheduleRehydration(vm);
+            }
+        }
+    });
+}
+
+function rehydrateHotStyle(style: StylesheetFactory) {
+    const list = activeStyles.get(style);
+    list?.forEach((vm) => {
+        if (!isUndefined(vm)) {
+            if (isFalse(vm.isDirty)) {
+                // hacky way to force the styles to get recomputed
+                // by replacing the value of old template, which is used
+                // during the rendering process. If the template returned
+                // by render() is different from the previous stored template
+                // the styles will be reset, along with the content of the
+                // shadow, this way we can guarantee that the styles will be
+                // recalculated, and applied.
+                vm.cmpTemplate = () => [];
+                // forcing the vm to rehydrate in the next tick
+                markComponentAsDirty(vm);
+                scheduleRehydration(vm);
+            }
+        }
+    });
+}
+
+function rehydrateHotComponent(Ctor: ComponentConstructor) {
+    const list = activeComponents.get(Ctor);
+    list?.forEach((vm) => {
+        if (!isUndefined(vm)) {
+            const { owner } = vm;
+            // the hot swapping for components only work for instances of components
+            // created from a template, root elements can't be swapped because we
+            // don't have a way to force the creation of the element with the same state
+            // of the current element.
+            if (!isNull(owner) && isFalse(owner.isDirty)) {
```

Comment:
Vue 3 automatically reloads the page if the root element is modified: https://github.com/vuejs/vue-next/blob/master/packages/runtime-core/src/hmr.ts#L108-L118.
I think it makes sense to do that automatically or to add a flag `reloadIfNeeded` or something like that to `registerComponentSwap` to reload the page if the component can't be swapped.

Reply:
@caridy said roo might be tricky, I think hot swapping the root doesn't look like a terrible corner case to have.

Reply:
@diervo what @pmdartus is saying is to reload the entire page (which we can definitely do, if the VM doesn't have an owner, which means it is a root element. I can do that for sure.

Reply:
Another approach would be to notify the caller that the swap can't be done and it's to the caller to decide what to do: display a banner on the page with a reload button, log a warning or an error, ...

Reply:
this is more complicated, because it is not a boolean operation:
a) the swap was possible (on at least 1 instance)
b) the swap was not possible in X active instances
c) the swap is not needed (no associated instance)

We could make it true/false if we say:

* true is problem was not encounter, which is a and c
* false if at least one instance was root (which mean we were not able to swap it)

Reply:
ok, figured a way for this that works fine... the apis are going to return a boolean, whether or not the operation succeed or not..., in the case of the template and style they will (for now) always return true. And for components, it will return false if a root instance is changing, so @diervo code can refresh the entire page